### PR TITLE
find agent_pid even when using a PEP397 launcher

### LINKED ISSUE #24361: [Core] ray.exceptions.LocalRayletDiedError: The task's local raylet died. Check raylet.out for more information.
### What happened + What you expected to happen

I was using ray as a dependency to run the Modin library in  Win 11 however I am getting Raylet error

### Versions / Dependencies

OS: Windows 11
Python: 3.7
Ray Version: 1.12.0

Ray agent failed
[dashboard.log](https://github.com/ray-project/ray/files/8596197/dashboard.log)
[dashboard_agent.log](https://github.com/ray-project/ray/files/8596198/dashboard_agent.log)

### Reproduction script

```
df.head(20):

Unnamed: 0                       game score home_odds draw_odds away_odds country                 league             datetime
0            0  Sport Recife - Imperatriz   2:2      1.36      4.31      7.66  Brazil  Copa do Nordeste 2020  2020-02-07 00:00:00
1            1           ABC - America RN   2:1      2.62      3.30      2.48  Brazil  Copa do Nordeste 2020  2020-02-02 22:00:00
2            2  Frei Paulistano - Nautico   0:2      5.19      3.58      1.62  Brazil  Copa do Nordeste 2020  2020-02-02 00:00:00
3            3    Botafogo PB - Confianca   1:1      2.06      3.16       3.5  Brazil  Copa do Nordeste 2020  2020-02-02 22:00:00
4            4          Fortaleza - Ceara   1:1      2.19      2.98      3.38  Brazil  Copa do Nordeste 2020  2020-02-02 22:00:00
5            5             River-PI - CSA   3:1      3.41      3.07       2.1  Brazil  Copa do Nordeste 2020  2020-02-02 20:00:00
6            6     Sport Recife - Vitoria   1:1      1.92      3.20      3.93  Brazil  Copa do Nordeste 2020  2020-02-02 20:00:00
7            7           CRB - Santa Cruz   1:0      2.22      3.06      3.18  Brazil  Copa do Nordeste 2020  2020-01-29 23:30:00
8            8         Bahia - Imperatriz   2:0      1.31      4.75      8.55  Brazil  Copa do Nordeste 2020  2020-01-29 00:00:00
9            9    Ceara - Frei Paulistano   2:2       1.2      5.43     12.94  Brazil  Copa do Nordeste 2020  2020-01-26 22:00:00
10          10            Confianca - ABC   1:0      2.09      3.01      3.58  Brazil  Copa do Nordeste 2020  2020-01-26 00:00:00
11          11   America RN - Botafogo PB   0:0      2.14      3.30      3.11  Brazil  Copa do Nordeste 2020  2020-01-26 22:00:00
12          12         CSA - Sport Recife   0:1      2.63      3.04      2.62  Brazil  Copa do Nordeste 2020  2020-01-26 22:00:00
13          13         Santa Cruz - Bahia   0:0      3.72      3.22      1.96  Brazil  Copa do Nordeste 2020  2020-01-26 20:00:00
14          14        Vitoria - Fortaleza   0:0      3.13      3.02      2.27  Brazil  Copa do Nordeste 2020  2020-01-26 20:00:00
15          15         Nautico - River-PI   1:1      1.44      4.06      6.19  Brazil  Copa do Nordeste 2020  2020-01-23 23:30:00
16          16           Imperatriz - CRB   2:1      2.34      2.94      3.08  Brazil  Copa do Nordeste 2020  2020-01-21 23:00:00
17          17       Nautico - Campinense   2:0       1.4      4.18      7.26  Brazil  Copa do Nordeste 2020  2019-05-16 00:30:00
18          18          CRB - Juazeirense   2:1       1.4      4.19      7.32  Brazil  Copa do Nordeste 2020  2019-05-14 23:00:00
19          19             ABC - AE Altos   2:0      1.44      3.83      7.02  Brazil  Copa do Nordeste 2020  2019-05-08 22:30:00
```

```
import os
import ray
ray.init()
os.environ["MODIN_ENGINE"] = "ray"
import modin.pandas as pd
df = pd.read_csv(r"data.csv", low_memory=False)
df['game'] = df['game'].astype(str).str.replace('(\(\w+\))', '', regex=True)
df['league'] = df['league'].astype(str).str.replace('(\s\d+\S\d+)$', '', regex=True)
df['league'] = df['league'].astype(str).str.replace('é', 'e', regex=True)
df['game'] = df['game'].astype(str).str.replace('(\s\d+\S\d+)$', '', regex=True)
df[['home_team', 'away_team']] = df['game'].str.split(' - ', expand=True, n=1)
df[['home_score', 'away_score']] = df['score'].str.split(':', expand=True)
df['away_score'] = df['away_score'].astype(str).str.replace('[a-zA-Z\s\D]', '', regex=True)
print (df.head())
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by mattip**
## Why are these changes needed?
When using ray inside a virtualenv on windows, `python.exe` as reported by `sys.executable` is a PEP397 launcher to the actual python as reported by `os.getpid()`:
```
>>> import sys, os, psutil
>>> >>> print(sys.executable)
C:\temp\issue24361\Scripts\python.exe
>>> os.getpid()
2208
>>> child = psutil.Process(2208)
>>> child.cmdline()
['C:\\oss\\CPython38\\python.exe']
>>> child.parent().cmdline()
['C:\\temp\\issue24361\\Scripts\\python.exe']
>>> child.parent().pid
6424
```

When the `agent_manager` launches the agent process via `Process::Process()`, it gets the PID of the launcher process (6424), which is what is expected as an ID when registering the agent in the gRPC callback. But inside `agent.py`, the child process reports the PID via `os.getpid()`, which is 2208, and this is the wrong PID to register the agent.

The solution proposed here checks whether `os.getpid()`'s parent is a python process, and if so assumes it is the launcher. It then sets this PID as the `agent_pid`. A different solution would be to pass the agent_pid in to the process as a command line parameter or an environment variable.

## Related issue number
Closes #24361, #23945

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

Is there a test for running the ray head process in a virtualenv (different than the runtime_env tests, which run the workers in a virtualenv/conda env)?

### CODE REVIEW

File: dashboard/agent.py [hunk 1]
```diff
@@ -200,10 +200,20 @@ async def _check_parent():
...
             self.aiogrpc_raylet_channel
         )
-
+        agent_pid = os.getpid()
+        parent = psutil.Process(agent_pid).parent()
+        # In a virtualenv, the agent PID could be the PEP397 launcher and
+        # not the python process PID as reported by os.getpid()
+        try:
+            pid = parent.pid
+        except AttributeError:
```
[the remark this answers is not in the document]
Comment (mattip) [reply dashboard/agent.py:212]: done

**Comment by mattip** [comment]
> Is there any reason why we shouldn't go with the alternative approach?

There are many ID types in `src/ray/common/id.h`. I am not sure how to best use one of them to create an AgentID in [AgentManager::StartAgent](https://github.com/ray-project/ray/blob/88dfc20c6e0b6d57794cdb565f221aa99c0b4882/src/ray/raylet/agent_manager.cc#L52) that would be used instead of the PID. Could you suggest some code to do this?

Once we have a unique identifier (hopefully not too long), it should be trivial to add a command-line argument to use it and then parse that argument in `agent.py`

**Comment by rkooo567** [comment]
The `Process` class, GetId should return the pid of the child processes!

**Comment by mattip** [comment]
closing in favor of #24968

**Pull request closed without merging**